File: sami/step_definitions/Loyalty Program.js
import { Given, When, Then } from 'cypress-cucumber-preprocessor/steps';
import LoyaltyProgramPage from '../pageObjects/LoyaltyProgramPage';

const loyaltyProgramPage = new LoyaltyProgramPage();

Given('the user is logged into their account', () => {
  // Implement login functionality here
  cy.login();
});

Given('the user is on the User Account page', () => {
  loyaltyProgramPage.visitUserAccountPage();
});

When('the user opts to join the Loyalty Program', () => {
  loyaltyProgramPage.joinLoyaltyProgram();
});

Then('the user should be successfully enrolled in the Loyalty Program', () => {
  loyaltyProgramPage.verifyEnrollmentSuccess();
});

Then('the user should receive a confirmation message', () => {
  loyaltyProgramPage.verifyConfirmationMessage();
});

Given('the user is enrolled in the Loyalty Program', () => {
  // Assume user is already enrolled
});

When('the user makes a purchase', () => {
  loyaltyProgramPage.makePurchase();
});

Then('the user should earn loyalty points for the purchase', () => {
  loyaltyProgramPage.verifyPointsEarned();
});

Then('the points should be reflected in the user's account', () => {
  loyaltyProgramPage.verifyPointsInAccount();
});

Given('the user has accumulated loyalty points', () => {
  // Assume user has points
});

When('the user chooses to redeem points for a discount', () => {
  loyaltyProgramPage.redeemPointsForDiscount();
});

Then('the discount should be applied to the user's purchase', () => {
  loyaltyProgramPage.verifyDiscountApplied();
});

Then('the redeemed points should be deducted from the user's account', () => {
  loyaltyProgramPage.verifyPointsDeducted();
});

When('the user navigates to the Loyalty section of their account', () => {
  loyaltyProgramPage.navigateToLoyaltySection();
});

Then('the user should see their current loyalty points balance', () => {
  loyaltyProgramPage.verifyPointsBalance();
});

When('there are new rewards or discounts available', () => {
  loyaltyProgramPage.checkNewRewardsOrDiscounts();
});

Then('the user should receive a notification about the new offers', () => {
  loyaltyProgramPage.verifyNotificationReceived();
});

When('the user opts to leave the Loyalty Program', () => {
  loyaltyProgramPage.leaveLoyaltyProgram();
});

Then('the user should be successfully unenrolled from the Loyalty Program', () => {
  loyaltyProgramPage.verifyUnenrollmentSuccess();
});

File: step_definitions/Loyalty Program.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class LoyaltyProgramSteps extends LoyaltyProgramPage {

    private TestHarness testHarness = new TestHarness();

    @Given("a customer is not enrolled in the loyalty program")
    public void customerIsNotEnrolled() {
        navigateToEnrollmentPage();
    }

    @When("the customer enrolls in the loyalty program")
    public void customerEnrolls() {
        enrollInLoyaltyProgram();
    }

    @Then("the customer should receive a confirmation of enrollment")
    public void confirmEnrollment() {
        verifyEnrollmentConfirmation();
    }

    @Then("the customer should see the loyalty program details in their account")
    public void verifyProgramDetails() {
        checkLoyaltyProgramDetails();
    }

    @Given("a customer is enrolled in the loyalty program")
    public void customerIsEnrolled() {
        verifyCustomerEnrollment();
    }

    @Given("the customer is logged into their account")
    public void customerIsLoggedIn() {
        verifyCustomerLogin();
    }

    @When("the customer makes a purchase")
    public void customerMakesPurchase() {
        makePurchase();
    }

    @Then("the customer should earn loyalty points for the purchase")
    public void earnLoyaltyPoints() {
        verifyLoyaltyPointsEarned();
    }

    @Then("the customer should receive any applicable discounts")
    public void receiveDiscounts() {
        verifyDiscountsApplied();
    }

    @Given("a customer has sufficient loyalty points")
    public void customerHasSufficientPoints() {
        verifySufficientLoyaltyPoints();
    }

    @When("the customer chooses to redeem points for a reward")
    public void redeemPointsForReward() {
        redeemLoyaltyPoints();
    }

    @Then("the loyalty points should be deducted from the customer's account")
    public void deductLoyaltyPoints() {
        verifyPointsDeduction();
    }

    @Then("the customer should receive the chosen reward")
    public void receiveChosenReward() {
        verifyRewardReceived();
    }

    @Given("a customer has made frequent purchases")
    public void customerHasFrequentPurchases() {
        verifyFrequentPurchases();
    }

    @When("the customer makes another purchase")
    public void customerMakesAnotherPurchase() {
        makeAnotherPurchase();
    }

    @Then("the customer should receive a discount based on their purchase frequency")
    public void receiveFrequencyDiscount() {
        verifyFrequencyDiscount();
    }

    @When("the customer checks their loyalty points balance")
    public void checkLoyaltyPointsBalance() {
        verifyLoyaltyPointsBalance();
    }

    @Then("the customer should see the correct number of loyalty points in their account")
    public void verifyCorrectPointsBalance() {
        checkCorrectPointsBalance();
    }

    @When("there is an update to the loyalty program")
    public void loyaltyProgramUpdate() {
        checkForProgramUpdates();
    }

    @Then("the customer should receive a notification about the update")
    public void receiveUpdateNotification() {
        verifyUpdateNotification();
    }
}